### ISSUE TITLE
relationsJoins enabled with @db.Json in Postgres makes the order of the JSON fields completely disappear

### ISSUE THREAD

**Issue by semoal** [prisma/prisma#22479]
### Bug description

We have a field called "layout" that is a JSON stored inside a template schema, we have noticed that our templates were crashing because out of order, after analyzing what happened we have saw that in the database was in the correct order but after calling:
```ts
const collectionWithTemplates = await prisma.collection.findMany({ include: { template: true }})
```

The template object which contains a `layout` field that is a `@db.JSON` was not ordered/preserved the order correctly.

### How to reproduce

Simply create an schema with a relation with json field, and fetch this field through a relation

### Expected behavior

It should keep the order

### Prisma schema

```prisma
generator client {
  provider        = "prisma-client-js"
  output          = "../../../packages/prisma-client/client"
  previewFeatures = ["fullTextSearch", "tracing", "relationJoins", "nativeDistinct"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Collection {
  id String @id @default(uuid()) @db.Uuid
  templateId String             @db.Uuid
  template   Template @relation(fields: [templateId], references: [id], onDelete: Cascade)
}

model Template {
  id String @id @default(uuid()) @db.Uuid
  layout       Json           @db.Json
  collections  Collection[]
}
```
### Code to check
```ts
  const template = await prisma.collection.findFirst({
    include: {
      template: true,
    },
  });
```

### Environment & setup

- OS: 
- Database: 
- Node.js version: 

Node 18, Ubuntu/Mac, PostgreSQL

### Prisma Version

```
5.7.1
```

**Comment by imwexpex** [comment]
We're facing the same issue which is critical for us. 

Anything I could help to confirm/solve it faster?

**Comment by renatocron** [comment]
Is this changing only the order of the keys, right? not the order of arrays, for example.

Relying on the order of keys inside a JSON is a little sketchy, I feel sorry for you

**Comment by imwexpex** [comment]
> Is this changing only the order of the keys, right? not the order of arrays, for example. If this is the case, Prisma (or rather, the JSONB) will remove any duplicate field, but that would happen anyway when loading then in a plain JS object
> 
> Relying on the order of keys inside a JSON is a little sketchy, I feel sorry for you.

What do you mean by `order of arrays`?

**Comment by renatocron** [comment]
```SQL
insert into "Template" (id, layout) select '8ba4599b-1116-4a14-bebc-1bfc0d14169a', '[1,2,3,4]';
insert into "Collection" (id, "templateId") select '00000000-1116-4a14-bebc-1bfc0d14169a', '8ba4599b-1116-4a14-bebc-1bfc0d14169a';
```

Prisma runs:
```SQL
SELECT "t1"."id", "t1"."templateId", "Collection_template"."__prisma_data__" AS "template" FROM "public"."Collection" AS "t1" LEFT JOIN LATERAL (SELECT COALESCE(JSONB_AGG("__prisma_data__"), '[]') AS "__prisma_data__" FROM (SELECT "t4"."__prisma_data__" FROM (SELECT JSONB_BUILD_OBJECT('id', "t3"."id", 'layout', "t3"."layout") AS "__prisma_data__" FROM (SELECT "t2".* FROM "public"."Template" AS "t2" WHERE "t1"."templateId" = "t2"."id" /* root select */) AS "t3" /* inner select */) AS "t4" /* middle select */) AS "t5" /* outer select */) AS "Collection_template" ON true LIMIT $1 [1] took 1ms
```

Order is preserved, as data is saved inside an array:
```js
{
  id: '00000000-1116-4a14-bebc-1bfc0d14169a',
  templateId: '8ba4599b-1116-4a14-bebc-1bfc0d14169a',
  template: {
    id: '8ba4599b-1116-4a14-bebc-1bfc0d14169a',
    layout: [ 1, 2, 3, 4 ]
  }
}
```

If we update the record to the typical key/value:

```SQL
update "Template" set layout = '{"first_key": "1", "second_key":"2", "...":"value"}'；
```

Prisma runs the same query:

```SQL
SELECT "t1"."id", "t1"."templateId", "Collection_template"."__prisma_data__" AS "template" FROM "public"."Collection" AS "t1" LEFT JOIN LATERAL (SELECT COALESCE(JSONB_AGG("__prisma_data__"), '[]') AS "__prisma_data__" FROM (SELECT "t4"."__prisma_data__" FROM (SELECT JSONB_BUILD_OBJECT('id', "t3"."id", 'layout', "t3"."layout") AS "__prisma_data__" FROM (SELECT "t2".* FROM "public"."Template" AS "t2" WHERE "t1"."templateId" = "t2"."id" /* root select */) AS "t3" /* inner select */) AS "t4" /* middle select */) AS "t5" /* outer select */) AS "Collection_template" ON true LIMIT $1 [1] took 0ms
```

```js
{
  id: '00000000-1116-4a14-bebc-1bfc0d14169a',
  templateId: '8ba4599b-1116-4a14-bebc-1bfc0d14169a',
  template: {
    id: '8ba4599b-1116-4a14-bebc-1bfc0d14169a',
    layout: { '...': 'value', first_key: '1', second_key: '2' }
  }
}
```
But the order is not preserved.

Running the query on the database, the order is already destroyed by Postgres, as you said before, because of the JSONB:

```txt
-[ RECORD 1 ]-----------------------------------------------------------------------------------------------------------------
id         | 00000000-1116-4a14-bebc-1bfc0d14169a
templateId | 8ba4599b-1116-4a14-bebc-1bfc0d14169a
template   | [{"id": "8ba4599b-1116-4a14-bebc-1bfc0d14169a", "layout": {"...": "value", "first_key": "1", "second_key": "2"}}]
```

JSON fields are stored as a string, but JSONB is stored as binary and is a bit optimized as well.

If I change the query to use JSON instead of JSONB, the order of the keys is preserved:

```SQL
SELECT "t1"."id", "t1"."templateId", "Collection_template"."__prisma_data__" AS "template" FROM "public"."Collection" AS "t1" LEFT JOIN LATERAL (SELECT COALESCE(JSON_AGG("__prisma_data__"), '[]') AS "__prisma_data__" FROM (SELECT "t4"."__prisma_data__" FROM (SELECT JSON_BUILD_OBJECT('id', "t3"."id", 'layout', "t3"."layout") AS "__prisma_data__" FROM (SELECT "t2".* FROM "public"."Template" AS "t2" WHERE "t1"."templateId" = "t2"."id" /* root select */) AS "t3" /* inner select */) AS "t4" /* middle select */) AS "t5" /* outer select */) AS "Collection_template" ON true LIMIT 1;
```

```txt
id         | 00000000-1116-4a14-bebc-1bfc0d14169a
templateId | 8ba4599b-1116-4a14-bebc-1bfc0d14169a
template   | [{"id" : "8ba4599b-1116-4a14-bebc-1bfc0d14169a", "layout" : {"first_key": "1", "second_key":"2", "...":"value"}}]
```

> In some cases, it's needed. The only option here is an additional round-trip for restoring original keys ordering in JS, for example using Zod schema

I'm still out of creativity in how the order of keys were persisted/restored into a dictionary should influence how the data is used. This behavior that other software is depending on is extremely fragile. That's why I said I was sorry; I'm not trying to be rude or anything.


Some languages even force that between each program start the hashes be randomized to prevent some hash-collision attacks, so it's kind of impossible to have the order preserved when reading JSON into an object except if implementing a custom reader.

According to the JSON specification (RFC 8259), the order of keys within an object is not guaranteed to be preserved. JSON is designed as a data interchange format, and parsers are free to rearrange the order of keys during processing.

To fix this, if the app is in him domain, he should use arrays or other data structures where order is explicitly maintained (per RFC, like any array in any software)

If the change in the struct is not possible, include an additional field to store the order explicitly if necessary

**Comment by Weakky** [comment]
Hey @imwexpex,

> In some cases, it's needed.

Would you mind clarifying what is your specific use case?

Thank you all for chiming in.

**Comment by imwexpex** [comment]
> Hey @imwexpex,
> 
> > In some cases, it's needed.
> 
> Would you mind clarifying what is your specific use case?
> 
> Thank you all for chiming in.

We're storing configuration for entry in JSON, where key ordering is valuable and directly affects business logic.
For example, draft order `{QB: 1, WR: 2}` => This means that 1 QB must be picked first, and 2 WRs second

**Comment by xentobias** [comment]
Is there any chance this can be fixed? This is currently blocking me from using the `relationJoins` feature in Prisma.

My use case is that I store the integration definition for a workflow editor as JSON in the database. When using `relationJoins`, the order of the pins changes and no longer matches the original declaration.

**Comment by semoal** [comment]
> Is there any chance this can be fixed? This is currently blocking me from using the `relationJoins` feature in Prisma.
> 
> My use case is that I store the integration definition for a workflow editor as JSON in the database. When using `relationJoins`, the order of the pins changes and no longer matches the original declaration.

You can pass a custom relation load at query level, we're using relation joins with the faster enabled, and for this joins with jsonb we're using the other

**Comment by xentobias** [comment]
Oh, that’s nice. Thank you for the hint! 👍

**Comment by renatocron** [comment]
I want to reiterate something more explicitly here, because this keeps coming up as if it were a Prisma bug.

@xentobias there's no chance this will be fixed without using postgres JSON with fake varchar in every op, to preserve the JSON as text, that would be slower.

This is **not really a Prisma issue**. The behavior comes from **PostgreSQL `jsonb` semantics**, which *explicitly do not preserve object key order*. Prisma is correctly using `jsonb` functions (`JSONB_BUILD_OBJECT`, `JSONB_AGG`) when `@db.Json` is backed by `jsonb`, and Postgres is free to reorder keys as part of normalization and optimization.

Relying on **object key order in JSON for business logic is a design flaw**:

* RFC 8259 does **not guarantee key ordering** in JSON objects
* PostgreSQL `jsonb` intentionally discards ordering
* Many runtimes and parsers do not guarantee stable ordering (some even randomize hashes)
* Any system depending on this is inherently fragile

If ordering is semantically meaningful, the data model should reflect that:

* Use **arrays** where order matters (this is exactly what arrays are for)
* Or store an **explicit order field**
* Or encode ordered data as a list of `{ key, order, value }`
* Or, if absolutely unavoidable, use `json` (not `jsonb`) with full awareness of the trade-offs (not possible with the aggregations the prisma does)

Expecting Prisma (or Postgres) to preserve JSON object key order would mean fighting both the database and the JSON spec itself. This isn’t something an ORM should “fix”, because doing so would give a false sense of correctness while remaining fundamentally unsafe.

In short: **JSON objects should never be treated as ordered structures**. If order matters, the schema needs to change.

**Comment by ak-lilw** [comment]
Yes, the Postgres `jsonb` type discards key order, but the `json` type explicitly [preserves it](https://www.postgresql.org/docs/current/datatype-json.html):

> Because the `json` type stores an exact copy of the input text, it will preserve [...] the order of keys within JSON objects.

Likewise, "many runtimes and parsers" might not guarantee object key order, but the ECMAScript standard itself [does](https://tc39.es/ecma262/#sec-ordinaryownpropertykeys) since ES2015:

> For each own property key `propertyKey` of `obj` such that `propertyKey` is a String and `propertyKey` is not an array index, in ascending chronological order of property creation [...]

RFC 8259 is a red herring; it only defines the textual syntax of JSON, not the meaning of values represented in that format. Sure, it doesn't say that objects should be parsed into a data structure which retains their order, but [it doesn't say that about arrays, either](https://datatracker.ietf.org/doc/html/rfc8259#section-5)!

So a schema which makes use of the Postgres `json` type (intentionally using `json` over `jsonb`) for one of its documented features, in an app targeting ES2015 or later, is relying strictly on guaranteed behaviour - it is not "inherently fragile" as you claim. It only breaks in this situation because Prisma does a conversion to `jsonb`, which is documented by Postgres to be a lossy conversion, but Prisma does not document that it may perform this lossy conversion when the user hasn't requested it.

In short, there is absolutely nothing wrong with using JSON objects to represent objects with ordered keys, so long as one is careful to use tools (such as Postgres and JavaScript) which offer guaranteed object key order. As Prisma is an interface between two such tools, it would be reasonable to expect that it doesn't break the guarantee made by both. Otherwise, Prisma should at least document that it has limited support for `json` field types in Postgres (e.g. [here](https://www.prisma.io/docs/orm/reference/prisma-schema-reference#postgresql-7)).

It is worth noting that the lossy conversion to `jsonb` can be observed in JavaScript in at least one other way, besides object key order. A `json` string may contain the escape sequence `\x0000`, but a `jsonb` string cannot since it is represented in the unescaped form.

**Comment by renatocron** [comment]
@ak-lilw 
Fair points on the `json` vs `jsonb` distinction and on ES2015 object ordering, 
I'll concede those, and the undocumented `json` => `jsonb` downcast in relationJoins is a legitimate Prisma gap worth documenting, but it's already prob only in the legacy driver, u prob just open a new for the prisma 8 where it's expected to be only used in Javascript, as the original prisma ideia was to be a framework across many langs.

Where I'd push back is the framing. Object key order being preserved is a property of the *serialization/decoding layer*, not of the JSON data model itself. It only holds when every hop happens to be in the order-preserving column, and plenty aren't:

- Rust `serde_json` default deserializes into a `BTreeMap`, so keys come out **alphabetically sorted** (you have to opt into `preserve_order`)
- Go `encoding/json` into `map[string]any` loses order entirely, maps randomize by design.
- C++ `nlohmann::json` default (`std::map`) sorts; you need `ordered_json`.
- Python only guarantees it since 3.7; Java `org.json.JSONObject` is `HashMap`-backed.

Compare with formats that take ordering seriously: Avro and Protobuf keep field order in the **schema**, and use `array`/`repeated` for ordered data. 

Protobuf `map<K,V>` explicitly leaves iteration/wire order undefined. The order that survives is the order expressed *structurally*, never the incidental key order of an associative container.

In the end, the code is yours to have, if your producer, store and consumer all stay in the order-preserving lane, it works. But IMHO that's an infra guarantee you're relying across every hop (and lot's of libs/authors in the middle, I think it's mainly fragile), not something the JSON object model gives you. If order is part of the logic, an array or an explicit index field encodes it safely regardless of runtime.
